### PR TITLE
Add Glibc 2.37

### LINKED ISSUE #9614: ChromeOS R123 updated to glibc 2.37, which breaks `crew`
Probably related: #9612

```text
chronos@maple / $ /lib/libc.so.6 
GNU C Library (Gentoo 2.37-r5 p5) stable release version 2.37.
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 10.2.0.
libc ABIs: UNIQUE ABSOLUTE
Minimum supported kernel: 3.2.0
For bug reporting instructions, please see:
<http://crbug.com/new>.

```

### PR DESCRIPTION
**Pull request by satmandu**
Fixes #9614 #9612 #9663
- Updates `rubocop`. (This is needed because the older version of `rubocop` errors out on the glibc package file.)
- Updates `install.sh` to add the conditions for using the `glibc_lib237` package on install.

Works properly:
- [x] `x86_64`
- [x] `armv7l` 

### Run the following to get this pull request's changes locally for testing.
```
CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=master crew update
```

### CODE REVIEW

**Comment by yest5xjquvf2** [comment]
not all packages are intel or acorn, amd exists

**Comment by satmandu** [comment]
Please let me know if the package I built works on an AMD system.

If not, then please submit a package file to build that binary. We will not accept user submissions of pre-built binaries. It's too much of a security risk.

**Comment by yest5xjquvf2** [comment]
WORKS!
Screenshot 2024-04-16 23 45 19

File: install.sh [hunk 1]
```diff
@@ -165,12 +165,19 @@ BOOTSTRAP_PACKAGES='zstd crew_mvdir ruby git ca_certificates libyaml openssl'
...
+if [[ -n "${CHROMEOS_RELEASE_CHROME_MILESTONE}" ]]; then
+  if (( "${CHROMEOS_RELEASE_CHROME_MILESTONE}" > "112" )); then
+    # Recent Arm systems have a cut down system.
+    [[ "${ARCH}" == "armv7l" ]] && BOOTSTRAP_PACKAGES+=' bzip2 ncurses readline pcre2 gcc_lib'
+  fi
 
-  # Recent Arm systems have a cut down system.
-  [[ "${ARCH}" == "armv7l" ]] && BOOTSTRAP_PACKAGES+=' bzip2 ncurses readline pcre2 gcc_lib'
+  if (( "${CHROMEOS_RELEASE_CHROME_MILESTONE}" < "123" )); then
+    # Append the correct packages for systems running M113 onwards.
```
Comment (uberhacker) [inline install.sh:175]: `Append the correct packages for systems running M122 and lower.`?

File: packages/glibc_build237.rb [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+          -I. -DDATADIR=\\\"#{CREW_PREFIX}/share\\\" -DPREFIX=\\\"#{CREW_PREFIX}\\\" \
+          -L../glibc_build \
+          -B../glibc_build/csu/ -lc -lc_nonshared"
+      end
+    end
+  end
+
+  def self.install
+    FileUtils.mkdir_p CREW_DEST_LIB_PREFIX
+    system "sed 's,/usr/#{ARCH_LIB}/libc_nonshared.a,#{CREW_LIB_PREFIX}/libc_nonshared.a,g' /usr/#{ARCH_LIB}/libc.so > #{CREW_DEST_LIB_PREFIX}/libc.so"
```
Comment (uberhacker) [inline packages/glibc_build237.rb:162]: Add to `def self.patch` section?

File: packages/glibc_build237.rb [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        system "make -j1 DESTDIR=#{CREW_DEST_DIR} install || make -j1 DESTDIR=#{CREW_DEST_DIR} install || true"
+      end
+      if @libc_version.to_f >= 2.32
+        system "install -Dt #{CREW_DEST_PREFIX}/bin -m755 build-locale-archive"
+        system "make -j1 DESTDIR=#{CREW_DEST_DIR} localedata/install-locales"
+      end
+      Dir.chdir CREW_DEST_LIB_PREFIX do
+        puts "System glibc version is #{LIBC_VERSION}.".lightblue
+        puts 'Creating symlinks to system glibc version to prevent breakage.'.lightblue
+        @crew_libc_version = @libc_version
```
Comment (uberhacker) [inline packages/glibc_build237.rb:179]: Why assign another variable?

File: packages/glibc_build237.rb [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+            elsif @opt_verbose
+              puts "#{f} excluded because #{@filetype}"
+            end
+          end
+        end
+      end
+    end
+    # Only save libnsl.so.2, since libnsl.so.1 is provided by perl
+    # For this to work, build on a M107 or newer container.
+    FileUtils.cp File.realpath("#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"), "#{CREW_DEST_LIB_PREFIX}/libnsl.so.2"
```
Comment (uberhacker) [inline packages/glibc_build237.rb:220]: Aren't you already indicating the full path?

**Review by uberhacker — CHANGES_REQUESTED** [review]
See comments.

File: packages/glibc_build237.rb [hunk 5]
```diff
@@ -0,0 +1,323 @@
+require 'package'
+
+class Glibc_build237 < Package
+  description 'The GNU C Library project provides the core libraries for GNU/Linux systems.'
+  homepage 'https://www.gnu.org/software/libc/'
+  license 'LGPL-2.1+, BSD, HPND, ISC, inner-net, rc, and PCRE'
+  compatibility 'all'
+  binary_compression 'tar.zst'
+  @libc_version = LIBC_VERSION
```
[the remark this answers is not in the document]
Comment (satmandu) [reply packages/glibc_build237.rb:9]: I leave it so it is easy to manually set when building on another Glibc version... For bootstrapping.

File: packages/glibc_build237.rb (hunk 2 above)
Comment (satmandu) [reply packages/glibc_build237.rb:162]: And then a FileUtils.install here?

File: packages/glibc_build237.rb (hunk 3 above)
Comment (satmandu) [reply packages/glibc_build237.rb:179]: Removed.

File: packages/glibc_build237.rb (hunk 4 above)
Comment (satmandu) [reply packages/glibc_build237.rb:220]: This. resolves symlinks.

File: install.sh (hunk 1 above)
Comment (uberhacker) [reply install.sh:175]: Why was this resolved?  Does the comment make any sense?

File: packages/glibc_build237.rb (hunk 2 above)
Comment (uberhacker) [reply packages/glibc_build237.rb:162]: Not sure.  Maybe you can patch the source prior to the build?

File: packages/glibc_build237.rb (hunk 4 above)
Comment (uberhacker) [reply packages/glibc_build237.rb:220]: Ah, that makes sense.

File: packages/glibc_build237.rb [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+            libc_cv_z_relro=yes \
+            --with-binutils=binutils \
+            --with-bugurl=https://github.com/chromebrew/chromebrew/issues/new \
+            --without-cvs \
+            --without-selinux \
+            "
+          # install-symbolic-link segfaults on armv7l, but we're deleting
+          # the libraries anyways, so it doesn't matter.
+          system "sed -i 's,install-symbolic-link,/bin/true,g' ../Makefile"
+          system "sed -i 's,symbolic-link-prog := $(elf-objpfx)sln,symbolic-link-prog := /bin/true,g' ../Makerules"
```
Comment (uberhacker) [inline packages/glibc_build237.rb:99]: Add to `def self.patch`?

File: install.sh (hunk 1 above)
Comment (satmandu) [reply install.sh]: The text was changed to `# Append the correct packages for systems running M122 and lower.`...

File: packages/glibc_build237.rb (hunk 6 above)
Comment (satmandu) [reply packages/glibc_build237.rb:99]: I don't think that works. These files are generated by autotools/configure...

File: install.sh (hunk 1 above)
Comment (uberhacker) [reply install.sh]: Sorry, I didn't refresh to see the updates. :blush:

File: packages/glibc_build237.rb (hunk 6 above)
Comment (uberhacker) [reply packages/glibc_build237.rb:99]: Fair enough.

File: packages/glibc_build237.rb (hunk 2 above)
Comment (satmandu) [reply packages/glibc_build237.rb:162]: Upon further reflection.... This is an intentional hack during the build process to let us work with the existing ChromeOS libraries... 

It's not really a patch, and it isn't really something where we can modify the source... We're just taking existing files and copying them over so we have them, with slight modifications so they refer to our paths...

**Review by uberhacker — APPROVED** [review]

**Pull request merged**